### ISSUE TITLE
we could provide an xml attribute text object

### ISSUE THREAD

**Issue by ninrod** [noctuid/targets.el#4]
for example

```html
<!-- cursor in [a]
command: dax -->
<a href="blah" cl[a]ss="bootclass">woot</a>

<!--result-->
<a href="blah">woot</a>
```

What do you think?

**Comment by ninrod** [comment]
In retrospect, an xml attribute seems a bit to specific for targets. targets seems more general: it will work on any filetype. xml attrs will only work in xml like files.

**Comment by noctuid** [comment]
I want to have targets be usable for any text object that it makes sense for. It's very easy to support anything that uses thing-at-pt.el. At some point, I'll probably add a mechanism to support other "types" of text objects if the user provides a select function and seek function for them, so even if I don't add this directly, it would be possible to create an xml text object with targets by using `evil-select-xml-tag` and function to seek for xml tags.

**Comment by ninrod** [comment]
That's interesting. Perhaps I could even work on that. Reopening...

**Comment by ninrod** [comment]
By the way, @noctuid, are you actually using targets.el for your day to day work? just curious, because I am.

**Comment by noctuid** [comment]
I am but mainly for testing purposes. The basic functionality works okay, but there are a lot of minor problems that I haven't had the time to fix.

**Comment by ninrod** [comment]
Maybe we could write a list of those problems somewhere in the repo. I'm looking forward to my next challenge on the targets.el package!

**Comment by ninrod** [comment]
hi @noctuid  

in #16 you've mentioned that you could provide information regarding the implementation of the xml text object functionality. It would mainly revolve around implementing a seeking function, you said.

Is `targets-seek-forward` a seeking function? In other words, if I were to implement the xml text object in targets.el, I should first write a `targets-xml-seek-forward` function and then add this to the `targets-seek-functions` list?

**Comment by noctuid** [comment]
I haven't exactly decided how I'm going to integrate it (see the last bullet under New Functionality in #16). I'm probably going to create some way for a user to specify a text object that is like one defined with `thing-at-point` (`evil-pargraph`, `evil-word`, etc. use `thing-at-point` by the way) but differentiates between `i` and `a`  in a different way. `evil-paragraph`, `evil-word`, etc. all exclude or include *whitespace* with `i` and `a`, but an xml tag text object behaves more like paren text objects, for example. You don't really need to worry about this. The seek function won't be added to `targets-seek-functions`; more likely, you'll end up doing something like `(put 'targets-xml-tag 'forward-op #'evil-xml-seek-forward)`.

For now, you would write a function that would seek forward to the beginning of an xml tag (e.g. `|<tag>...</tag>` where `|` is the point) and took `count` and `bound` arguments. The function should seek backwards if a negative count is given. You could write two separate functions for seeking forward and backward and combine them later if you want.

Seeking behavior is a little more complicated for things that can be nested. I'm going to allow for different behavior in nested cases in the future, but the seeking functions you create should ignore the nested case. By that I mean, that they should always jump to the beginning of the next tag that starts after the point. For example:
```
<tag1>...<tag2>...| </tag2>...</tag1>...<tag3></tag3>
seeking forward should jump here:
...|<tag3></tag3>
```
Similarly, seeking backward should seek to the start of the closest tag to the point. For example:
```
<tag1>...<tag2>... </tag2>...</tag1>...|<tag3></tag3>
seeking backward should jump here:
<tag1>...|<tag2>... </tag2>...</tag1>...<tag3></tag3>
```

**Bascially, you can ignore close tags and only jump between open tags.** Just to give an example of what alternative behavior would look like (using parens as the example):
```
|(a (b c) d) (e)
seek forward
(a |(b c) d) (e)
seek forward
(a (b c) d|) (e)
seek forward
(a (b c) d) |(e)
```

The initial seeking functions do not need to act like this. Let me know if you want me to clarify anything.

**Comment by ninrod** [comment]
hi @noctuid, I'm not sure the intended behaviour is clear. I think you are designing an xml tag text object, while I had in mind an xml *attribute* text object, like this:

```html
<!-- cursor in [a]
command: dax -->
<a href="bar" cl[a]ss="bootclass">foo</a>

<!--result-->
<a href="bar["]>foo</a>
```
I'm pretty sure that a "tag" textobj already exists by default. For example, I use `dat`, `vat` and `cit` all the time.

**Comment by noctuid** [comment]
Sorry, my bad. Targets does not support the tag text object currently though. To add an xml attribute text object, you would have to do everything from scratch (both the seeking and selection functions). How would `dix` behave?

**Comment by ninrod** [comment]
Ok, np. I think `dix` would preserve the space between xml attributes, similarly to `diw` vs `daw`. 
here's a [demo](https://github.com/whatyouhide/vim-textobj-xmlattr) on vim side.

**Comment by noctuid** [comment]
In that case, you could implement this without me making any changes to targets by using vanilla `thing-at-point`. You just need to provide the necessary functions. Afterwards you would do something like this:
```
(put 'targets-xml-attribute 'forward-op #'evil-xml-attribute-seek-forward)
(put 'targets-xml-attribute 'beg-op ...)
(put 'targets-xml-attribute 'end-op ...)
```

`beg-op` and `end-op` would move the point to the beginning of end of an attribute respectively and are used to get the selection. You can also instead provide a single selection function that returns the bounds as a cons (e.g. `(1 . 10)`) in which case you would do `(put 'targets-xml-attribute 'bounds-of-thing-at-point #'function-to-return-xml-attribute-bounds)`.

**Comment by ninrod** [comment]
nice, but I did not understand what do I have to do with the `'targets-xml-attribute` thing that I'm creating there? How do I tell targets: "look targets, here's a new text object I've defined. Make him play along with the rest of the text objects, will you?"

**Comment by noctuid** [comment]
Once you've defined a thing, you would define a text object for it with targets like `(targets-define-to targets-xml-attribute 'targets-xml-attribute nil object)`. You could add a definition to `targets-object-text-objects` (but don't bind it by default since it's not replacing a default text object). As long as the `thing-at-point` operations all work (`end-of-thing`, `beginning-of-thing`, `forward-thing`, etc.), no more information is needed.

**Comment by ninrod** [comment]
nice. I think I'll try to define a very dumb *thing* like trying to reimplement the `W` text object, which should be easy. Once I've integrated all the parts into targets.el, then I'll try to refactor the `W` thing into the `xml-attribute-thing`.

**Comment by noctuid** [comment]
Yeah, looking at the evil-WORD functions might be helpful as well. Evil doesn't explicitly use `put`, it instead names function in a way so that `thing-at-point` will recognize them (e.g. `forward-evil-WORD`).

**Comment by ninrod** [comment]
Hi @noctuid. I went out and implemented a text object for xml attributes which you may want to check out. I found it usable enough that people may want to use it standalone if they don't want all the funcionality provided by targets.el. Anyway we could try to integrate it here to, perhaps as a dependency, somehow. 

Here's the link: https://github.com/ninrod/exato

**Comment by noctuid** [comment]
If you implement the text object as a thing, targets can use it without any extra work. I added a more detailed explanation of using thingatpt/creating text objects [here](https://github.com/noctuid/targets.el#creating-newintegrating-existing-text-objects-using-thingatptel). It looks like you've already done most of the necessary work.

**Comment by ninrod** [comment]
ahh, excelent. this provides crucial extra information about  the interplay between `thing-at-point` and `evil`. Good job documenting it.
